File: leetcode/tree/2773_Height_of_Special_Binary_Tree/codes/dfs.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def heightOfTree(self, root: Optional[TreeNode]) -> int:
        count_map = {}

        def dfs(node):
            if not node:
                return
            count_map[node.val] = count_map.get(node.val, 0) + 1
            if count_map[node.val] > 1:
                return
            dfs(node.left)
            dfs(node.right)

        dfs(root)

        def max_depth(node, depth):
            if not node or count_map[node.val] > 1:
                return depth
            left = max_depth(node.left, depth + 1)
            right = max_depth(node.right, depth + 1)
            return max(left, right)

        return max_depth(root, 0)

File: leetcode/tree/2773_Height_of_Special_Binary_Tree/codes/dfs.cpp
/**
Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int heightOfTree(TreeNode* root) {
        unordered_map<int, int> countMap;
        dfs(root, countMap);
        return maxDepth(root, 0, countMap);
    }

private:
    void dfs(TreeNode* node, unordered_map<int, int>& countMap) {
        if (!node) return;
        countMap[node -> val]++;
        if (countMap[node -> val] > 1) return;
        dfs(node -> left, countMap);
        dfs(node -> right, countMap);
    }

    int maxDepth(TreeNode* node, int depth, unordered_map<int, int>& countMap) {
        if (!node || countMap[node -> val] > 1) return depth;
        int left = maxDepth(node -> left, depth + 1, countMap);
        int right = maxDepth(node -> right, depth + 1, countMap);
        return max(left, right);
    }
};
